src/main/java/org/apache/mesos/scheduler/MesosToSchedulerDriverAdapter.java

```diff
@@ -0,0 +1,731 @@
+    private boolean registered;
+    private Mesos mesos;
+    private volatile Timer subscriberTimer;
+    private State state;
+    private Timer heartbeatTimer;
+    private Instant lastHeartbeat;
+    private OptionalLong heartbeatTimeout;
+
+    public MesosToSchedulerDriverAdapter(org.apache.mesos.Scheduler wrappedScheduler,
+                                         org.apache.mesos.Protos.FrameworkInfo frameworkInfo,
+                                         String master) {
+        this.wrappedScheduler = wrappedScheduler;
+        this.frameworkInfo = Evolver.evolve(frameworkInfo);
+        this.credential = null;
+        this.master = master;
+        this.registered = false;
+        this.status = org.apache.mesos.Protos.Status.DRIVER_NOT_STARTED;
+        this.state = State.DISCONNECTED;
+    }
+
+    public MesosToSchedulerDriverAdapter(org.apache.mesos.Scheduler wrappedScheduler,
+                                         org.apache.mesos.Protos.FrameworkInfo frameworkInfo,
+                                         String master,
+                                         org.apache.mesos.Protos.Credential credential) {
+        this.wrappedScheduler = wrappedScheduler;
+        this.frameworkInfo = Evolver.evolve(frameworkInfo);
+        this.master = master;
+        this.credential = Evolver.evolve(credential);
+        this.registered = false;
+        this.status = org.apache.mesos.Protos.Status.DRIVER_NOT_STARTED;
+        this.state = State.DISCONNECTED;
+    }
+
+    @Override
+    public synchronized void connected(Mesos mesos) {
+        if (status != org.apache.mesos.Protos.Status.DRIVER_RUNNING) {
+            return;
+        }
+
+        LOGGER.info("Connected!");
+
+        state = State.CONNECTED;
+
+        performReliableSubscription();
+    }
+
+    private synchronized void subscribe(ExponentialBackOff backOff) throws IOException {
+        if (state == State.SUBSCRIBED || state == State.DISCONNECTED) {
+            LOGGER.debug("Cancelling subscriber. As we are in state: {}", state);
+            cancelSubscriber();
+            return;
+        }
+        LOGGER.info("Sending SUBSCRIBE call");
+
+        final Protos.Call.Builder callBuilder = Protos.Call.newBuilder()
+                .setType(Protos.Call.Type.SUBSCRIBE)
+                .setSubscribe(Protos.Call.Subscribe.newBuilder()
+                        .setFrameworkInfo(frameworkInfo)
+                        .build());
+        if (frameworkInfo.hasId()) {
+            callBuilder.setFrameworkId(frameworkInfo.getId());
+        }
+
+        mesos.send(callBuilder.build());
+
+        scheduleNextSubscription(backOff);
+    }
+
+    private synchronized void scheduleNextSubscription(ExponentialBackOff backOff) throws IOException {
+        long nextBackoffMs;
+        nextBackoffMs = backOff.nextBackOffMillis();
+        LOGGER.info("Backing off for: {}", nextBackoffMs);
+        subscriberTimer.schedule(new SubscriberTask(backOff), nextBackoffMs);
```

Comment:
We can, let me try. Using `SchedulerExecutorService` certainly has many advantages over `Timer`.